audioutils/fmsynth/fmsynth.c

```diff
@@ -0,0 +1,238 @@
+    {
+      out += fmsynthop_operate(op, snd->phase_time);
+    }
+
+  update_phase(snd);
+
+  return out * snd->volume / FMSYNTH_MAX_VOLUME;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * name: fmsynth_initialize
+ ****************************************************************************/
+
+int fmsynth_initialize(int fs)
+{
+  max_phase_time = fs * WRAP_ROUND_TIME_SEC;
+  return fmsynthop_set_samplerate(fs);
+}
+
+/****************************************************************************
+ * name: fmsynthsnd_create
+ ****************************************************************************/
+
+FAR fmsynth_sound_t *fmsynthsnd_create(void)
+{
+  FAR fmsynth_sound_t *ret;
+  ret = (FAR fmsynth_sound_t *)malloc(sizeof(fmsynth_sound_t));
+  if (ret)
+    {
+      ret->phase_time = 0;
+      ret->volume = FMSYNTH_MAX_VOLUME;
+      ret->operators = NULL;
+      ret->next_sound = NULL;
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * name: fmsynthsnd_delete
+ ****************************************************************************/
+
+void fmsynthsnd_delete(FAR fmsynth_sound_t *snd)
+{
+  if (snd)
+    {
+      free(snd);
+    }
+}
+
+/****************************************************************************
+ * name: fmsynthsnd_set_operator
+ ****************************************************************************/
+
+int fmsynthsnd_set_operator(FAR fmsynth_sound_t *snd, FAR fmsynth_op_t *op)
+{
+  snd->operators = op;
+
+  return OK;
+}
+
+/****************************************************************************
+ * name: fmsynthsnd_set_soundfreq
+ ****************************************************************************/
+
+void fmsynthsnd_set_soundfreq(FAR fmsynth_sound_t *snd, float freq)
+{
+  FAR fmsynth_op_t *op;
+
+  for (op = snd->operators; op; op = op->parallelop)
+    {
+      fmsynthop_set_soundfreq(op, freq);
+      fmsynthop_start(op);
+    }
+}
+
+/****************************************************************************
+ * name: fmsynthsnd_set_volume
+ ****************************************************************************/
+
+void fmsynthsnd_set_volume(FAR fmsynth_sound_t *snd, float vol)
+{
+  snd->volume = vol * FMSYNTH_MAX_VOLUME;
+}
+
+/****************************************************************************
+ * name: fmsynthsnd_add_subsound
+ ****************************************************************************/
+
+int fmsynthsnd_add_subsound(FAR fmsynth_sound_t *top,
+                            FAR fmsynth_sound_t *sub)
+{
+  FAR fmsynth_sound_t *s = top;
+
+  if (!top || !sub)
+    {
+      return ERROR;
+    }
+
+  for (s = top; s->next_sound; s = s->next_sound);
+  s->next_sound = sub;
```

Comment:
Do we need `sub->next_sound = NULL;`?

Reply:
This is intentional, because I want to just connect sounds and mix each.

Reply:
ok